src/test/java/htsjdk/samtools/util/IOUtilTest.java

```diff
@@ -520,4 +527,253 @@ public void testIsBlockCompressedOnJimfs(Path file, boolean checkExtension, bool
+        File tmpDir = new File(tmpPath, userName);
+        tmpDir.mkdir();
+        tmpDir.deleteOnExit();
+        for(int cl=1;cl<=9;cl++)
+        {
+            Path outFile=Paths.get(tmpDir.getAbsolutePath(),"tmp"+extension);
+            IOUtil.setCompressionLevel(cl);
+            Assert.assertEquals(IOUtil.getCompressionLevel(),cl);
+            InputStream inStream=IOUtil.openFileForReading(file);
+            OutputStream outStream=IOUtil.openFileForWriting(outFile.toFile());
+            IOUtil.transferByStream(inStream,outStream,origSize);
+            outStream.close();
+            outFile.toFile().deleteOnExit();
+            long newSize=outFile.toFile().length();
+            if(cl<=lastDifference)Assert.assertTrue(previousSize>newSize);
+            else Assert.assertTrue(previousSize>=newSize);
+            previousSize=newSize;
+
+        }
+
+    }
+
+    @DataProvider
+    public static Object [][] badCompressionLevels(){
+        return new Object[][]{
+                {-1},
+                {10}
+        };
+
+    }
+
+    @Test(dataProvider = "badCompressionLevels")
+    public void testCompressionLevelExceptions(int cl){
+        Assert.assertThrows(IllegalArgumentException.class,()->IOUtil.setCompressionLevel(cl));
+    }
+
+    @DataProvider
+    public static Object [][] filesToCopy() {
+        return new Object[][]{
+                {TEST_VARIANT_DIR.resolve("test1.vcf")},
+                {TEST_DATA_DIR.resolve("ipsum.txt")}
+
+        };
+    }
+
+    @Test(dataProvider = "filesToCopy")
+    public void testCopyFile(Path file) throws IOException{
+        String tmpPath = System.getProperty("java.io.tmpdir");
+        String userName=System.getProperty("user.name");
+        File tmpDir = new File(tmpPath, userName);
+        tmpDir.mkdir();
+        tmpDir.deleteOnExit();
+        Path outFile=Paths.get(tmpDir.getAbsolutePath(),"tmp"+file.getFileName());
+        IOUtil.copyFile(file.toFile(),outFile.toFile());
+
+        outFile.toFile().deleteOnExit();
+        Stream<String> stream=Files.lines(file);
+        Stream<String> outStream=Files.lines(outFile);
+
+        Iterator<String> itStream=stream.iterator();
+        Iterator<String> itOutStream=outStream.iterator();
+
+        while(itStream.hasNext() && itOutStream.hasNext()){
+            Assert.assertEquals(itStream.next(),itOutStream.next());
+        }
+        Assert.assertFalse(itStream.hasNext());
+        Assert.assertFalse(itOutStream.hasNext());
+    }
+
+    @Test(dataProvider="filesToCopy")
+    public void testCopyFileException(Path file){
+        String tmpPath = System.getProperty("java.io.tmpdir");
+        String userName=System.getProperty("user.name");
+        File tmpDir = new File(tmpPath, userName);
+        tmpDir.mkdir();
+        tmpDir.deleteOnExit();
+        Path outFile=Paths.get(tmpPath,userName,"tmp"+file.getFileName());
+        ProcessExecutor.execute(new String[]{"touch", outFile.toAbsolutePath().toString()});

```

Comment:
I think it would be better to open and close the file than to spawn an external command here.